Report fixes (#70)
SW709E19/Report: report/PartDesign/ChapterPrototype/Prototype.tex

Through a short initial analysis it was concluded that MQTT was the optimal fit as the communication protocol in terms of the prototype. 
Other options such as Bluetooth Low Energy was however considered, but was deemed too time consuming to implement as of the current state of development and the benefits would only be in relation to battery life which was not of concern in this prototype.

\section{Gateway}
The gateway consists of a Raspberry Pi 3 Model B+ \cite{RaspberryPi3} which serves as a local Wi-Fi hotspot such that the sensor devices can be connected and their data be transferred.
The gateway is only responsible for passing on the devices' data to the cloud in this prototype. 

> The communication between sensor devices, gateway, and cloud was managed by using Mosquitto \cite{Mosquitto}, a message broker for the MQTT protocol which can be configured to bridge topics to a remote broker.  

Reviewer: messages instead of topics?

> The communication between sensor devices, gateway, and cloud was managed by using Mosquitto \cite{Mosquitto}, a message broker for the MQTT protocol which can be configured to bridge topics to a remote broker.  

Reviewer 2: ```suggestion
The communication between sensor devices, gateway, and cloud was managed by using Mosquitto \cite{Mosquitto}, a message broker for the MQTT protocol which can be configured to bridge messages to a remote broker.  
```

\section{Cloud architecture}
The cloud provider chosen was Google Cloud Platform (GCP) since Neurospace ApS were willing to provide us with full access to their company account. 
The minimized setup time and monetary advantage of using GCP therefore exceeded the need for exploration of alternative cloud providers.

> The main concern of the cloud computing platform used for the prototype was to expose an endpoint which the gateway could publish data for persistent storage.

Reviewer: I think he wanted us to replace computing with platform "cloud computing" -> "cloud platform"

\end{figure}

The first service, IoT Core \cite{IoT-Core}, is a fully managed platform for managing IoT devices. 
The service provides multiple useful features such as device management through registration and authentication as well as two-way secure communication between cloud and devices/gateways. 

> The service natively supports MQTT, which made it relatively straightforward to set up communication with the gateway. This was done by configuring the gateway to bridge some of the topics to IoT Core's MQTT endpoint.

Reviewer: Again, I dont think topics are the right terminology. And why only "some"

Reviewer: Maybe it makes sense... Not sure

Reviewer 2: ```suggestion
The service natively supports MQTT, which made it relatively straightforward to set up communication with the gateway.
This was done by configuring the gateway to bridge some of the topics to IoT Core's MQTT endpoint.
```
Topics bør være korrekt da vi specifisere nogle bestemte topics der skal subscribes på. Alle beskeder der kommer på dette topic vil så blive sendt videre.

The messages received from the gateways are automatically transferred to another GCP service: Pub/Sub.
The Pub/Sub \cite{PubSub} service is a fully managed event ingestion and delivery system which makes messages available in buckets\footnote{A bucket is a storage component for services on Google Cloud Platform.} for subscribers.
In order to later inspect messages from devices, the Pub/Sub was configured to trigger a function that writes all device events into an event log accessible through Google Cloud's management tool, Stackdriver \cite{google-stackdriver}.

\section{Limitations}

> Before the gateway can publish and subscribe to messages on behalf of a device, the device must be attached to the gateway.

Reviewer: subscribe to messages? topics?

> Before the gateway can publish and subscribe to messages on behalf of a device, the device must be attached to the gateway.

Reviewer 2: ```suggestion
Before the gateway can publish messages and subscribe to topics on behalf of a device, the device must be attached to the gateway.
```

The gateway attaches a device by publishing a message to IoT Core's MQTT endpoint at topic: /devices/\{device-id\}/attach \cite{IoTCoreAttach}.
The gateway can then publish and subscribe on behalf of the device until it detaches the device.
It was not possible to configure Mosquitto to automatically attach devices before subscribing to device topics from IoT Core.
This meant that the prototype was not able to support messages from the cloud to the devices, including configuration updates.

> Therefore, another solution for bridging the gateway and IoT Core must be implemented to support configuration updates.

Reviewer: "bridging the gateway and IoT Core". Does this make sense?

> Therefore, another solution for bridging the gateway and IoT Core must be implemented to support configuration updates.

Reviewer 2: ```suggestion
Therefore, another solution for bridging the devices/gateways and cloud must be implemented to support configuration updates.
```

\section{Results}

The development of the prototype resulted in knowledge about several areas of interest for the final solution:
\begin{itemize}
	\item How to configure the ESP32 to read measurements from a sensor.
	\item The sensor used to measure vibrations could measure at a sample rate of about 300 Hz.
	\item Transmitting messages from the sensor device to the gateway using Wi-Fi consumes a lot of power. Therefore, to minimise power consumption the device should transmit messages as seldom as possible.
	\item How to send data wirelessly from the sensor device to the gateway and from the gateway to the cloud using MQTT.
	\item The Google Cloud services used were convenient for the prototype, but resulted in a solution which depends on services provided by Google Cloud. Furthermore, the price of using the services may become an issue if the number of messages sent to the cloud exceed certain bounds.

> 	\item Another solution for bridging the gateway and IoT Core must be implemented to support configuration updates.

Reviewer: "bridging the gateway and IoT Core"

> 	\item Another solution for bridging the gateway and IoT Core must be implemented to support configuration updates.

Reviewer 2: ```suggestion
	\item Another solution for bridging the devices/gateways and cloud must be implemented to support configuration updates.
```

\end{figure}

The first service, IoT Core \cite{IoT-Core}, is a fully managed platform for managing IoT devices. 
The service provides multiple useful features such as device management through registration and authentication as well as two-way secure communication between cloud and devices/gateways. 
The service natively supports MQTT, which made it relatively straightforward to set up communication with the gateway.

> This was done by configuring the gateway to bridge some of the topics to IoT Core's MQTT endpoint.

Reviewer: ```suggestion
This was done by configuring the gateway to bridge the topics to IoT Core's MQTT endpoint.
```
